plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/revisited/ModulesManager.java

```diff
@@ -596,6 +598,8 @@ public static void buildKeysForRegularEntries(IProgressMonitor monitor, ModulesF
 
         int j = 0;
         FastStringBuffer buffer = new FastStringBuffer();
+        Set<Map.Entry<File, String>> packageEntries = new HashSet<Map.Entry<File, String>>();
```

Comment:
Seeing it now, I think this should be a Map<String, File> and not a set... as key entries you just need the package names (so, you'll have to iterate less the 2nd time).

So, instead of putting the whole path there, just put `FullRepIterable.getParentModule(m)` as the key and the file with the last value (you'll end up re-adding a given entry multiple times -- that's ok). Also, you don't need to skip the `__init__` here, adding it is ok as the validation will be done later on anyways, so, there's no need to check.